anchore_engine/services/policy_engine/engine/vulns/cpes.py

```diff
@@ -0,0 +1,181 @@
+import re
+import typing
+from typing import List, Tuple
+
+from anchore_engine.db.entities.policy_engine import ImageCpe, ImagePackage
+from anchore_engine.common import os_package_types
+
+
+def compare_fields(lhs, rhs):
+    """
+    Comparison function for cpe fields for ordering
+    - * is considered least specific and any non-* value is considered greater
+    - if both sides are non-*, comparison defaults to python lexicographic comparison for consistency
+    """
+
+    if lhs == "*":
+        if rhs == "*":
+            return 0
+        else:
+            return 1
+    else:
+        if rhs == "*":
+            return -1
+        else:
+            # case where both are not *, i.e. some value. pick a way to compare them, using lexicographic comparison for now
+            if rhs == lhs:
+                return 0
+            elif rhs > lhs:
+                return 1
+            else:
+                return -1
+
+
+def compare_cpes(lhs: ImageCpe, rhs: ImageCpe):
+    """
+    Compares the cpes based on business logic and returns -1, 0 or 1 if the lhs is lower than, equal to or greater than the rhs respectively
+
+    Business logic here is to compare vendor, name, version, update and meta fields in that order
+    """
+    vendor_cmp = compare_fields(lhs.vendor, rhs.vendor)
+    if vendor_cmp != 0:
+        return vendor_cmp
+
+    name_cmp = compare_fields(lhs.name, rhs.name)
+    if name_cmp != 0:
+        return name_cmp
+
+    version_cmp = compare_fields(lhs.version, rhs.version)
+    if version_cmp != 0:
+        return version_cmp
+
+    update_cmp = compare_fields(lhs.update, rhs.update)
+    if update_cmp != 0:
+        return update_cmp
+
+    meta_cmp = compare_fields(lhs.meta, rhs.meta)
+    if meta_cmp != 0:
+        return meta_cmp
+
+    # all avenues of comparison have been depleted, the two cpes are same for all practical purposes
+    return 0
+
+
+class FuzzyCandidateCpeGenerator:
+    """
+    A generator for CPEs from package metadata that generates candidates predictively
+    """
+
+    embedded_semver_regex = re.compile(r".*([0-9]+\.[0-9]+\.[0-9]+).*")
+
+    def for_distro_package(self, package: ImagePackage) -> typing.List[ImageCpe]:
+        """
+        Create the cpes for a single distro package
+
+        :param package:
+        :return: list of ImageCpes for the package
+        """
+        cpes = []
+
+        # Do a "-" --> "_" substitution addition so may have multiple cpe candidates for a single package
+        names = {package.name, re.sub("-", "_", package.name)}
+        if not package.pkg_path and package.pkg_type in os_package_types:
+            pkg_path = "pkgdb"
+        else:
+            pkg_path = package.pkg_path
+
+        for name in names:
+            # for vendor in vendors:
+            c = ImageCpe()
+            c.name = name
+            c.version = package.version
+            c.pkg_type = package.pkg_type
+            c.cpetype = "a"
```

Comment:
Just curious, what does the "a" mean?

Reply:
In cpes there are "h" - hardware, "o" - OS, and "a" - applications